FpsCheck.py

```diff
@@ -366,33 +364,48 @@ def find_invalid_well_known(self, check_sets):
-                                field_sym_difference.update(
-                                    set(json_schema[field][aliased_site]) ^ 
-                                    set(
-                                    curr_fps_set.relevant_fields_dict[field]
-                                    [aliased_site]))
-                    if field_sym_difference:
-                        self.error_list.append("The following member(s) of " 
-                        + field + " were not present in both the changelist "
-                        + "and .well-known/first-party-set.json file: " + 
-                        str(sorted(field_sym_difference)))
+                well_known_set = FpsSet(
+                    json_schema.get('ccTLDs'), 
+                    json_schema.get('primary'), 
+                    json_schema.get('associatedSites'), 
+                    json_schema.get('serviceSites'))
+                if well_known_set.primary != curr_fps_set.primary:
+                    field_sym_difference = [well_known_set.primary, curr_fps_set.primary]
+                    self.error_list.append("The following primary was not "
+                    + "present in both the changelist and "
+                    + ".well-known/first-party-set.json file: "
+                    + str(sorted(field_sym_difference)))
+                if well_known_set.associated_sites != curr_fps_set.associated_sites:
+                    field_sym_difference = set(
+                       well_known_set.associated_sites) ^ set(curr_fps_set.associated_sites)
+                    self.error_list.append("The following member(s) of " 
+                    + "associatedSites were not present in both the "
+                    + "changelist and .well-known/first-party-set.json "
+                    + "file: " + str(sorted(field_sym_difference)))
+                if well_known_set.service_sites != curr_fps_set.service_sites:
+                    field_sym_difference = set(
+                       well_known_set.service_sites) ^ set(curr_fps_set.service_sites)
+                    self.error_list.append("The following member(s) of " 
+                    + "serviceSites were not present in both the "
+                    + "changelist and .well-known/first-party-set.json "
+                    + "file: " + str(sorted(field_sym_difference)))
+                if well_known_set.ccTLDs != curr_fps_set.ccTLDs:
+                    field_sym_difference = set(
+                        well_known_set.ccTLDS.keys()) ^ set(
+                            curr_fps_set.ccTLDs.keys())
+                    self.error_list.append("The following aliased ccTLD sites"
+                    + "were not present in both the changelist and " 
+                    + ".well-known/first-party-set.json file: "
+                    + str(sorted(field_sym_difference)))
+                    for aliased_site in curr_fps_set.ccTLDs & well_known_set.ccTLDs:
```

Comment:
Does this work? AFAIK the `ccTLDs` field is a dict and dicts don't have a `&` operator.

Aside from that - we want to iterate over the union (not the intersection) of the sets of keys, right?

Reply:
I believe they do have the & operator ([an example](https://www.geeksforgeeks.org/python-intersect-two-dictionaries-through-keys/#:~:text=Method%20%232%3A%20Using%20%26%20operator)). I did want to use the intersection here, idea being that we don't want to have a key error when trying to access any site that's aliased on only the well-known or the PR submission. Any `{aliased_site: alias}` pair in which the `aliased_site` is not a key in both the well-known ccTLDs and the submission ccTLDs is automatically considered to be a well-known error without looking at their values, which is taken care of in the [previous step](https://github.com/GoogleChrome/first-party-sets/blob/6c40c62c4a87cc88a02d1b770ab481ead2f39305/FpsCheck.py#L393C21-L395C56). Perhaps the error message there could be altered to include both the sites and their aliases?

Reply:
That example uses the `&` operator on `ini_dict1.items()` and `ini_dict2.items()`, not on the dictionaries themselves. There is no `&` operator for dicts:
```python
>>> import sys
>>> sys.version
'3.11.5 (main, Aug 25 2023, 07:43:52) [GCC 12.2.0]'
>>> d = {'x': 'y'}
>>> d & d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for &: 'dict' and 'dict'
>>> d.items() & d.items()
{('x', 'y')}
```

(The supported operators for dicts are listed below [here](https://docs.python.org/3/library/stdtypes.html#dict.values).)

> I did want to use the intersection here ...

Aha, I see. In that case I think it would be simpler to loop over the union of the keys, and use `aliased_site in dict` or `dict.get(aliased_site, [])` instead of `dict[aliased_site]` to ensure that we don't get a KeyError? That way we could handle all of the aliased sites the same way, instead of handling them differently depending on the shape of the two dicts.

Reply:
Ah, sorry, you're right. I had gotten it mixed up with the `|` operator. 
>> we could handle all of the aliased sites the same way, instead of handling them differently depending on the shape of the two dicts

I'm not sure I fully understand how this would look - if a key is present in one dict and not the other, we would have to treat the process a bit differently regardless as there wouldn't be two values to take the diff of. I suppose incorporating the order matters approach here would alleviate that problem a bit, but the results could overall be less useful. We can talk about this offline to clarify a bit.